src/uct/cuda/base/cuda_md.c

```diff
@@ -55,6 +56,62 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_detect_memory_type,
     return UCS_ERR_INVALID_ADDR;
 }
 
+static ucs_status_t uct_cuda_base_get_sys_dev(ucs_sys_device_t *sys_dev_p)
+{
+    CUdevice dev;
+    int attrib;
+
+    if (UCS_OK != UCT_CUDADRV_FUNC(cuCtxGetDevice(&dev))) {
```

Comment:
why is the memory address ignored?

Reply:
To populate device information, we need to get the device associated (`CUdevice dev`) with current GPU context. The expectation is that when a call leading to `md_mem_query` is made, the context is already set -- viz a safe assumption because a buffer pointer has been passed. Once the context is set, it's possible to get GPU context associated with buffer using [cuPointerGetAttribute](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__UNIFIED.html#group__CUDA__UNIFIED_1g0c28ed0aff848042bc0533110e45820c) but AFAIK there is no way to get a `CUdevice` from a specific GPU context. Instead we have a call to get `CUdevice` for the currently set context using [cuCtxGetDevice](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1g4e84b109eba36cdaaade167f34ae881e) and in the process we end up ignoring the memory address to generate SYS_DEV info.

Reply:
seems problematic that the function which queries pointer properties ignores the pointer
what if the application changes GPU context during its lifetime?

Reply:
@Akshay-Venkatesh  what is device CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL attribute you get with  cuPointerGetAttribute?

Reply:
> seems problematic that the function which queries pointer properties ignores the pointer
what if the application changes GPU context during its lifetime?

@yosefe This isn't allowed in most runtimes today. Processes maintain a 1:1 mapping and use a single gpu context for the entirety of the process lifespan (at least under MPI and other regimes we know of). But the suggestion from @bureddy works. We can extract ordinal from pointer and then get device from the ordinal. Lemme make this change.

Reply:
ok. BTW, is it possible different threads in the process will use different GPUs?

Reply:
This is also not supported by any MPI runtime that I know of. This subject is pretty broad (a bit like MPI threads support) . This may not be exhaustive:
1. Single process using a single GPU
2. Single process using multiple GPUs from the same thread
3. Single thread from a process binding to 1 GPU
4. Single thread from a process binding to many GPUs
5. Other alltoall bindings...

The main reason for limited support being that we'd have to maintain additional state in UCX and we don't have users requesting this at least until now (arguable though).
AFAIK only NCCL permits 1, 2, 3 based on the [documentation](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/overview.html) which says:
> NCCL is compatible with virtually any multi-GPU parallelization model, for example:
>    single-threaded control of all GPUs
>    multi-threaded, for example, using one thread per GPU
>    multi-process, for example, MPI